File: code/cpp/ch22/learn/example_01_stack_queue_basics.cpp
/*
 * Example 01: Stack & Queue Basics — See LIFO and FIFO in Action
 * Chapter 22: Stacks & Queues — Order Matters
 *
 * Demonstrates:
 *   - Stack (LIFO) using std::stack
 *   - Queue (FIFO) using std::queue
 *   - Balanced parentheses with a stack
 *
 * Build: g++ -std=c++17 -o /tmp/ex01_ch22 code/cpp/ch22/learn/example_01_stack_queue_basics.cpp && /tmp/ex01_ch22
 */

#include <iostream>
#include <queue>
#include <stack>
#include <string>
#include <vector>
using namespace std;

void stackDemo() {
    cout << "=== Stack (LIFO) Demo ===" << endl;
    stack<int> stk;
    stk.push(10); cout << "push(10)" << endl;
    stk.push(20); cout << "push(20)" << endl;
    stk.push(30); cout << "push(30)" << endl;
    cout << "top() -> " << stk.top() << endl;
    stk.pop();
    cout << "pop() -> top is now " << stk.top() << endl;
    cout << "size: " << stk.size() << ", empty: " << stk.empty() << endl;
    cout << endl;
}

void queueDemo() {
    cout << "=== Queue (FIFO) Demo ===" << endl;
    queue<string> q;
    q.push("Alice");   cout << "enqueue Alice" << endl;
    q.push("Bob");     cout << "enqueue Bob" << endl;
    q.push("Charlie"); cout << "enqueue Charlie" << endl;
    cout << "front() -> " << q.front() << endl;
    q.pop();
    cout << "dequeue -> front is now " << q.front() << endl;
    cout << "size: " << q.size() << endl;
    cout << endl;
}

bool isValid(const string& s) {
    stack<char> stk;
    for (char ch : s) {
        if (ch == '(' || ch == '[' || ch == '{') {
            stk.push(ch);
        } else {
            if (stk.empty()) return false;
            char top = stk.top(); stk.pop();
            if ((ch == ')' && top != '(') ||
                (ch == ']' && top != '[') ||
                (ch == '}' && top != '{')) return false;
        }
    }
    return stk.empty();
}

void balancedParens() {
    cout << "=== Balanced Parentheses ===" << endl;
    vector<string> tests = {"({[]})", "([)]", "((()))", "(((", ""};
    for (const string& s : tests) {
        cout << "  \"" << s << "\" -> " << (isValid(s) ? "VALID" : "INVALID") << endl;
    }
}

int main() {
    stackDemo();
    queueDemo();
    balancedParens();
    return 0;
}

File: code/java/ch22/learn/Example01StackQueueBasics.java
package ch22.learn;

import java.util.*;

/**
 * Example 01: Stack & Queue Basics — See LIFO and FIFO in Action
 * Chapter 22: Stacks & Queues — Order Matters
 *
 * Demonstrates:
 *   - Stack (LIFO) using ArrayDeque
 *   - Queue (FIFO) using ArrayDeque
 *   - Balanced parentheses with a stack
 */
public class Example01StackQueueBasics {

    public static void stackDemo() {
        System.out.println("=== Stack (LIFO) Demo ===");
        Deque<Integer> stack = new ArrayDeque<>();
        stack.push(10); System.out.println("push(10) -> " + stack);
        stack.push(20); System.out.println("push(20) -> " + stack);
        stack.push(30); System.out.println("push(30) -> " + stack);
        System.out.println("peek()   -> " + stack.peek());
        System.out.println("pop()    -> " + stack.pop());
        System.out.println("After pop: " + stack);
        System.out.println();
    }

    public static void queueDemo() {
        System.out.println("=== Queue (FIFO) Demo ===");
        Queue<String> queue = new ArrayDeque<>();
        queue.offer("Alice");   System.out.println("enqueue Alice -> " + queue);
        queue.offer("Bob");     System.out.println("enqueue Bob   -> " + queue);
        queue.offer("Charlie"); System.out.println("enqueue Charlie -> " + queue);
        System.out.println("peek()   -> " + queue.peek());
        System.out.println("poll()   -> " + queue.poll());
        System.out.println("After dequeue: " + queue);
        System.out.println();
    }

    public static void balancedParens() {
        System.out.println("=== Balanced Parentheses ===");
        String[] tests = {"({[]})", "([)]", "((()))", "(((", ""};
        for (String s : tests) {
            boolean valid = isValid(s);
            System.out.println("  \"" + s + "\" -> " + (valid ? "VALID" : "INVALID"));
        }
    }

    static boolean isValid(String s) {
        Deque<Character> stack = new ArrayDeque<>();
        for (char ch : s.toCharArray()) {
            if (ch == '(' || ch == '[' || ch == '{') {
                stack.push(ch);
            } else {
                if (stack.isEmpty()) return false;
                char top = stack.pop();
                if ((ch == ')' && top != '(') ||
                    (ch == ']' && top != '[') ||
                    (ch == '}' && top != '{')) return false;
            }
        }
        return stack.isEmpty();
    }

    public static void main(String[] args) {
        stackDemo();
        queueDemo();
        balancedParens();
    }
}
